devp2p/kademlia.py

```diff
@@ -421,22 +425,22 @@ def _expected_pongs():
         if pingid and (pingid not in self._expected_pongs):
             assert pingid not in self._expected_pongs
             log.debug('surprising pong', remoteid=node,
-                      expected=_expected_pongs(), pingid=pingid.encode('hex')[:8])
+                      expected=_expected_pongs(), pingid=encode_hex(pingid)[:8])
             if pingid in self._deleted_pingids:
                 log.debug('surprising pong was deleted')
             else:
                 for key in self._expected_pongs:
                     if key.endswith(node.pubkey):
                         log.debug('waiting for ping from node, but echo mismatch', node=node,
-                                  expected_echo=key[:len(node.pubkey)][:8].encode('hex'),
-                                  received_echo=pingid[:len(node.pubkey)][:8].encode('hex'))
+                                  expected_echo=encode_hex(key[:len(node.pubkey)][:8]),
+                                  received_echo=encode_hex(pingid[:len(node.pubkey)][:8]))
             return
 
         # check for timed out pings and eventually evict them
-        for _pingid, (timeout, _node, replacement) in self._expected_pongs.items():
+        for _pingid, (timeout, _node, replacement) in list(self._expected_pongs.items()):
```

Comment:
In relation to the mutation I don't see a real problem in this specific case, because we're creating a list of the items to iterate over while deleting elements in original dict. Also the logging inside might be useful for tracking issues.